scipy/optimize/__slsqp.h

```diff
@@ -0,0 +1,405 @@
+    free(indices);
+    double* mem_ret = realloc(buffer, n*sizeof(double));
+    // Very unlikely, but just in case
+    if (mem_ret == NULL)
+    {
+        free(buffer);
+        PYERR(slsqp_error, "Memory reallocation failed.");
+    }
+    npy_intp shape_ret[1] = {n};
+    PyArrayObject* ap_ret = (PyArrayObject*)PyArray_SimpleNewFromData(1, shape_ret, NPY_FLOAT64, mem_ret);
+    // Return the result
+    return Py_BuildValue("Ndi",PyArray_Return(ap_ret), rnorm, info);
+
+}
+
+
+static PyObject*
+slsqp(PyObject* Py_UNUSED(dummy), PyObject* args)
+{
+    PyArrayObject *ap_gradx=NULL, *ap_C=NULL, *ap_d=NULL, *ap_mult=NULL;
+    PyArrayObject *ap_sol =NULL, *ap_xl=NULL, *ap_xu=NULL, *ap_buffer=NULL;
+    PyArrayObject* ap_indices=NULL;
+    PyObject* input_dict = NULL;
+    double funx;
+    struct SLSQP_static_vars Vars;
+
+    // The Python input should provide with a dictionary that maps to the struct
+    // SLSQP_static_vars. Necessary fields that would make the algorithm change
+    // behavior are m, meq, n, acc, maxiter, and mode. The rest can be left as zero.
+    // Changing values mid run is not recommended as they hold the internal state
+    // of the algorithm.
+    // The reason why they are returned is to make the algorithm stateless.
+
+    // The required arrays C, d, x, xl, xu, gradx, sol are passed as numpy arrays.
+    // The remaining arrays are going to be allocated in the buffer.
+
+    if (!PyArg_ParseTuple(args, "O!dO!O!O!O!O!O!O!O!O!",
+                          &PyDict_Type, (PyObject **)&input_dict,
+                          &funx,
+                          &PyArray_Type, (PyObject **)&ap_gradx,
+                          &PyArray_Type, (PyObject **)&ap_C,
+                          &PyArray_Type, (PyObject **)&ap_d,
+                          &PyArray_Type, (PyObject **)&ap_sol,
+                          &PyArray_Type, (PyObject **)&ap_mult,
+                          &PyArray_Type, (PyObject **)&ap_xl,
+                          &PyArray_Type, (PyObject **)&ap_xu,
+                          &PyArray_Type, (PyObject **)&ap_buffer,
+                          &PyArray_Type, (PyObject **)&ap_indices))
+    {
+        return NULL;
+    }
+
+    // Some helper x macros to pack and unpack the SLSQP_static_vars struct and
+    // the Python dictionary.
+
+    #define STRUCT_DOUBLE_FIELD_NAMES X(acc) X(alpha) X(f0) X(gs) X(h1) X(h2) X(h3) X(h4) X(t) X(t0) X(tol)
+    #define STRUCT_INT_FIELD_NAMES X(exact) X(inconsistent) X(reset) X(iter) X(itermax) X(line) X(m) X(meq) X(mode) X(n)
+    #define STRUCT_FIELD_NAMES STRUCT_INT_FIELD_NAMES STRUCT_DOUBLE_FIELD_NAMES
+
+    // Parse the dictionary, if the field is not found, raise an error.
+    // Do it separately for doubles and ints.
+    // Initialize the struct that will be populated from dict with zeros
+    #define X(name) Vars.name = 0;
+    STRUCT_FIELD_NAMES
+    #undef X
+
+    // PyDict_GetItemString returns a borrowed reference.
```

Comment:
Is this cargo culted from somewhere, or newly written? It looks a little funky. Probably because `SLSQP_static_vars` is a pretty unusual struct. Also, why the `static` in the name?

Reply:
I wrote it for maintenance reasons actually. But nice that it sounded like AI Slop hahah. 

Specifically, we are communicating with the Python side through a dict which holds the current state to make things thread-safe. So each time C code sends the control back to Python side we have to pack the struct into a Python dict. But CPython API is a bit fun in the sense that sometimes you have to ref count sometimes you don't. They call it strong and weak(borrowed) reference. So `GetItemString` is weak but `SetItemString` is strong so you need to involve ref counting in the X-macro.

So this is supposed to be a note about that particular detail.

Reply:
`static` in the sense that original code had these `SAVE`d hence they were not thread-safe in C `SAVE` is done via declaring them `static`. Now that I moved them into a dict that's what I meant but might be confusing indeed.

Reply:
So they're not actually declared `static`. Are you saying that the current version is still not thread-safe? I thought that it is thread-safe now. If so, it'd be good to remove `_static` from the name. If not, I think `pytest-run-parallel` should have caught it.

Reply:
No they are now thread-safe now that we do the dictionary ping-pong. The name is supposed to be saying `these are the variables that were previously static`. 

But yeah I already confused you meaning I'll confuse others too. Let me rename it.